ir/src/transition_constraints/graph.rs

```diff
@@ -187,9 +195,69 @@ impl AlgebraicGraph {
                 let node_index = self.insert_op(Operation::PeriodicColumn(index, cycle_len));
                 Ok((trace_segment, node_index))
             }
+            IdentifierType::Constant(ConstantType::Scalar(scalar_const)) => {
+                let trace_segment = 0;
+                let node_index = self.insert_op(Operation::ScalarConstant(scalar_const));
+                Ok((trace_segment, node_index))
+            }
             _ => Err(SemanticError::InvalidUsage(format!(
-                "Identifier {} was declared as a {} not as a trace column",
-                ident, col_type
+                "Identifier {} was declared as a {} which is not a supported type.",
+                ident, elem_type
+            ))),
+        }
+    }
+
+    /// Validates and adds a vector access to the graph.
+    /// Returns an error if the vector identifier's value is not of a supported type.'
+    fn insert_vector_access(
+        &mut self,
+        symbol_table: &SymbolTable,
+        vector_access: &VectorAccess,
+    ) -> Result<(TraceSegment, NodeIndex), SemanticError> {
+        // check that vector access is valid
+        symbol_table.validate_vector_access(vector_access.name(), vector_access.idx())?;
+        let vector_access_type = symbol_table.get_type(vector_access.name())?;
+        match vector_access_type {
+            IdentifierType::Constant(ConstantType::Vector(_)) => {
+                let trace_segment = 0;
+                let node_index = self.insert_op(Operation::VectorConstant(vector_access.clone()));
+                Ok((trace_segment, node_index))
+            }
+            _ => Err(SemanticError::InvalidUsage(format!(
+                "Vector Access {}[{}] was declared as a {} which is not a supported type.",
+                vector_access.name(),
+                vector_access.idx(),
+                vector_access_type
+            ))),
+        }
+    }
+
+    /// Validates and adds a matrix access to the graph.
+    /// Returns an error if the matrix identifier's value is not of a supported type.
+    fn insert_matrix_access(
```

Comment:
Similar comments here as for `insert_vector_access()`.

Reply:
Same as above. The signature now looks like:
```
pub(super) fn validate_matrix_access(
        &self,
        matrix_access: &MatrixAccess,
        matrix_access_type: &IdentifierType,
    ) -> Result<(), SemanticError>
```